ros/angel_system_nodes/angel_system_nodes/hmm.py

```diff
@@ -0,0 +1,254 @@
+from threading import Event, Thread
+import time
+
+import rclpy
+from rclpy.node import Node
+
+from angel_msgs.msg import (
+    ActivityDetection,
+    TaskUpdate,
+    TaskGraph,
+)
+from angel_msgs.srv import QueryTaskGraph
+from angel_system.activity_hmm.core import ActivityHMMRos
+from angel_utils.conversion import time_to_int
+
+
+class HMMNode(Node):
+    """
+    ROS node that runs the HMM and publishes TaskUpdate messages. The HMM is
+    called on a separate thread.
+    """
+    def __init__(self):
+        super().__init__(self.__class__.__name__)
+        log = self.get_logger()
+
+        self._det_topic = (
+            self.declare_parameter("det_topic", "ActivityDetections")
+            .get_parameter_value()
+            .string_value
+        )
+        self._config_file = (
+            self.declare_parameter("config_file",
+                                   "config/tasks/task_steps_config-recipe_coffee.yaml")
+            .get_parameter_value()
+            .string_value
+        )
+        self._task_state_topic = (
+            self.declare_parameter("task_state_topic", "TaskUpdates")
+            .get_parameter_value()
+            .string_value
+        )
+
+        # Instantiate the HMM module
+        self._hmm = ActivityHMMRos(self._config_file)
```

Comment:
I see that this is being used across different threads. I don't think I saw any thread safety in this object, so you likely need to create an associated lock to guard the usage of this resource throughout this node.